File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/dictionary/ExternalLookupURLBuilder.kt
// region Shared-Core Candidate
// Pure logic, Kotlin stdlib only (java.net.URLEncoder is JDK stdlib,
// accepted per the InputNormalizer / java.text.Normalizer precedent).
// Eligible for cross-platform extraction.
// endregion

package com.siansiansu.taigikeyboard.ime.dictionary

import com.siansiansu.taigikeyboard.engine.nfdPreprocessForLookup
import com.siansiansu.taigikeyboard.engine.RustEngineBridge
import com.siansiansu.taigikeyboard.engine.stripTone
import java.net.URLEncoder

/**
 * Builds MOE / Chhoe Taigi external dictionary lookup URLs from TL display form.
 * Parallels iOS `ExternalLookupURLBuilder.swift`. Tone-digit conversion semantics
 * follow the external dictionaries' URL format (tone 1 / tone 4 omitted).
 *
 * Platform-owned by design — URL conventions (which tones to drop, which
 * query parameters each dictionary takes, percent-encoding rules) are
 * not phonetics; they belong with the platform that knows about
 * `URLEncoder`. The phonetic prep step delegates to
 * `RustEngineBridge.nfdPreprocessForLookup` and tone-strip to
 * `RustEngineBridge.stripTone`.
 */
object ExternalLookupURLBuilder {
    /** Build Chhoe Taigi lookup URL. Returns null if the TL string produces an empty digit form. */
    fun chhoeURL(tl: String): String? {
        val tlDigit = toTLDigit(tl)
        if (tlDigit.isEmpty()) return null
        val encoded = URLEncoder.encode(tlDigit, "UTF-8")
        return "https://chhoe.taigi.info/s?s=su&f=e&lmjf=ki&lmj=$encoded"
    }

    /** Build MOE Dictionary lookup URL. Returns null if the TL string produces an empty digit form. */
    fun moeURL(tl: String): String? {
        val tlDigit = toTLDigit(tl)
        if (tlDigit.isEmpty()) return null
        val encoded = URLEncoder.encode(tlDigit, "UTF-8")
        return "https://sutian.moe.edu.tw/zh-hant/tshiau/?lui=tai_su&tsha=$encoded"
    }

    /** Convert TL display form (with diacritics) to TL digit form for URL. */
    fun toTLDigit(tl: String): String {
        val syllables = tl.lowercase().split("-")
        return syllables.joinToString("-") { normalizeSyllableToDigit(it) }
    }

    private fun normalizeSyllableToDigit(syllable: String): String {
        if (syllable.isEmpty()) return ""
        // Quick path: nasal-only substitution suffices to detect already-digit
        // input (e.g. "ho2", "saⁿ1"). Full POJ preprocessing (`o͘` → `oo`) is
        // only needed for the diacritic branch below.
        val withNasalConverted = syllable.replace("ⁿ", "nn").replace("ᴺ", "nn")
        if (withNasalConverted.last().isDigit()) {
            val tone = withNasalConverted.last().toString()
            if (tone == "1" || tone == "4") return withNasalConverted.dropLast(1)
            return withNasalConverted
        }
        // Diacritic path: apply full Taigi preprocessing (nasal + `o͘` → `oo`)
        // before tone stripping — matches iOS `ExternalLookupURLBuilder` and
        // ensures custom-dict entries containing POJ `o͘` produce canonical
        // TL `hoo`/`hoo2` URLs instead of `ho͘`/`ho͘2`.
        val preprocessed = RustEngineBridge.nfdPreprocessForLookup(syllable)
        val stripped = RustEngineBridge.stripTone(preprocessed)
        if (stripped.tone.isEmpty() || stripped.tone == "1" || stripped.tone == "4") {
            return stripped.bare
        }
        return stripped.bare + stripped.tone
    }
}

File: ios/Sources/TaigiKeyboard/Lexicon/Utils/ExternalLookupURLBuilder.swift
// 中文: 把 TL 顯示形(帶調符)轉成數字調並組成外部辭典查詢 URL。

import Foundation

/// Builds external dictionary lookup URLs from Taiwanese TL forms.
///
/// Third-party dictionaries (Chhoe Taigi, MOE) expect *digit-toned* TL
/// syllables (e.g. `tai7-tsi3`), while our search results carry the
/// diacritic display form (`tāi-tsì`). This helper encapsulates that
/// conversion and percent-encoding so the lookup call sites stay trivial.
///
/// Platform-owned by design — URL conventions (which tones to drop, which
/// query parameters each dictionary takes, percent-encoding rules) are
/// not phonetics; they belong with the platform that knows about
/// `URLQueryAllowed` / `URLEncoder`. The phonetic prep step delegates to
/// `RustEngineBridge.nfdPreprocessForLookup` and tone-strip to
/// `RustEngineBridge.stripTone`. Mirror at
/// `android/.../ime/dictionary/ExternalLookupURLBuilder.kt`.
// 中文: URL 慣例屬於平台層(percent-encoding / query 參數),所以這裡刻意不搬到 Rust;
// 中文: 只有調符前處理與 strip-tone 走 RustEngineBridge。
enum ExternalLookupURLBuilder {
    /// Chhoe Taigi dictionary lookup URL for the given TL display form.
    // 中文: Chhoe Taigi 辭典查詢 URL。
    static func chhoeURL(forTL tl: String) -> URL? {
        guard let encoded = encodedTLDigit(tl) else { return nil }
        return URL(string: "https://chhoe.taigi.info/s?s=su&f=e&lmjf=ki&lmj=\(encoded)")
    }

    /// MOE Sutian dictionary lookup URL for the given TL display form.
    // 中文: 教育部臺語辭典(Sutian)查詢 URL。
    static func moeURL(forTL tl: String) -> URL? {
        guard let encoded = encodedTLDigit(tl) else { return nil }
        return URL(string: "https://sutian.moe.edu.tw/zh-hant/tshiau/?lui=tai_su&tsha=\(encoded)")
    }

    /// Convert TL display form (diacritics) to TL digit form.
    /// e.g. `"tāi-tsì"` → `"tai7-tsi3"`.
    // 中文: 把 TL 顯示形轉成數字調形,例:tāi-tsì → tai7-tsi3。
    static func toTLDigit(_ tl: String) -> String {
        let syllables = tl.lowercased().split(separator: "-", omittingEmptySubsequences: false)
        return syllables
            .map { normalizeSyllableToDigit(String($0)) }
            .joined(separator: "-")
    }

    private static func encodedTLDigit(_ tl: String) -> String? {
        let digit = toTLDigit(tl)
        guard !digit.isEmpty else { return nil }
        return digit.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)
    }

    /// Normalize a single syllable from diacritics to digit tone.
    ///
    /// External dictionary URLs omit tones 1 (open) and 4 (checked), matching
    /// the query convention used by both Chhoe Taigi and MOE Sutian.
    // 中文: 單音節 diacritic → 數字調轉換。1 / 4 聲在 URL 中省略,
    // 中文: 與 Chhoe Taigi、MOE Sutian 的 query 慣例一致。
    private static func normalizeSyllableToDigit(_ syllable: String) -> String {
        guard !syllable.isEmpty else { return "" }

        // Quick path: already-digit-toned input keeps the digit (or strips
        // tone 1 / 4 for external dictionary URL semantics). Note: we
        // intentionally check the digit on the *raw* input — only the
        // nasal-marker substitution matters for the digit-strip branch,
        // and the full preprocessing happens below for the diacritic path.
        let withNasalConverted = syllable
            .replacingOccurrences(of: "\u{207F}", with: "nn")
            .replacingOccurrences(of: "\u{1D3A}", with: "nn")
        if let lastChar = withNasalConverted.last, lastChar.isNumber {
            let tone = String(lastChar)
            if tone == "1" || tone == "4" {
                return String(withNasalConverted.dropLast())
            }
            return withNasalConverted
        }

        // Diacritic path: apply Taigi preprocessing (nasal / o͘ normalization)
        // then reuse the shared tone-stripping helper so all call sites share
        // one implementation.
        let preprocessed = RustEngineBridge.nfdPreprocessForLookup(syllable)
        let stripped = RustEngineBridge.stripTone(preprocessed)
        let bare = stripped.bare
        let tone = stripped.tone

        // Tone 1 (open) and 4 (checked) are omitted in external dictionary URLs.
        if tone.isEmpty || tone == "1" || tone == "4" {
            return bare
        }
        return bare + tone
    }
}
